Purchases/Purchasing/StoreKit2/ProductsFetcherSK1.swift

```diff
@@ -0,0 +1,127 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  ProductsFetcherSK1.swift
+//
+//  Created by Andrés Boedo on 7/26/21.
+
+import Foundation
+import StoreKit
+
+class ProductsFetcherSK1: NSObject {
+    private let productsRequestFactory: ProductsRequestFactory
+
+    private var cachedProductsByIdentifier: [String: SKProduct] = [:]
+    private let queue = DispatchQueue(label: "ProductsFetcherSK1")
+    private var productsByRequests: [SKRequest: Set<String>] = [:]
+    private var completionHandlers: [Set<String>: [(Set<SKProduct>) -> Void]] = [:]
+
+    init(productsRequestFactory: ProductsRequestFactory = ProductsRequestFactory()) {
+        self.productsRequestFactory = productsRequestFactory
+    }
+
+    func products(withIdentifiers identifiers: Set<String>,
+                  completion: @escaping (Set<SKProduct>) -> Void) {
+
+        queue.async { [self] in
+            let productsAlreadyCached = self.cachedProductsByIdentifier.filter { key, _ in identifiers.contains(key) }
+            if productsAlreadyCached.count == identifiers.count {
+                let productsAlreadyCachedSet = Set(productsAlreadyCached.values)
+                Logger.debug(String(format: Strings.offering.products_already_cached, identifiers))
+                completion(productsAlreadyCachedSet)
+                return
+            }
+
+            if let existingHandlers = self.completionHandlers[identifiers] {
+                Logger.debug(String(format: Strings.offering.found_existing_product_request, identifiers))
+                self.completionHandlers[identifiers] = existingHandlers + [completion]
+                return
+            }
+
+            Logger.debug(
+                String(format: Strings.offering.no_cached_requests_and_products_starting_skproduct_request, identifiers)
+            )
+            let request = self.productsRequestFactory.request(productIdentifiers: identifiers)
+            request.delegate = self
+            self.completionHandlers[identifiers] = [completion]
+            self.productsByRequests[request] = identifiers
+            request.start()
+        }
+    }
+}
+
+extension ProductsFetcherSK1: SKProductsRequestDelegate {
+
+    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
+        queue.async { [self] in
```

Comment:
Can't we make this class an `actor` and do async await here? So we can remove this queue and make `completionHandlers` and `productsByRequests` thread safe?

I think the main change would be to make this function:

```
    func products(request: SKProductsRequest) async throws -> SKProductsResponse {
```

If it's too much of a change compared to the original version I understand, but I think we can take advantage of async await here

Reply:
I would love to do that... but actors are iOS 15+, so we can only safely use them for SK2 stuff

Reply:
got it!!! I keep getting confused with Xcode 13 requirement vs iOS 15 requirement. Sorry for the noise!

Reply:
no worries!!